File: Leetcode946/resolve.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
var validateStackSequences = function(pushed, popped) {
    var stk = Array();
    for (var i = 0, j = 0; j < popped.length; j++) {
        while (i < pushed.length && (stk.length == 0 || stk[stk.length - 1] != popped[j])) {
            stk.push(pushed[i]);
            i++;
        }
        if (stk.pop() != popped[j]) return false;
    }
    return true;
};

File: Leetcode946/resolve.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> stk;
        int push = 0, pop = 0;
        while (push < pushed.size()) {
            cout << pop << endl;
            while (push < pushed.size() && pushed[push] != popped[pop]) stk.push(pushed[push++]);
            pop++, push++;
            while (!stk.empty() && stk.top() == popped[pop]) {
                stk.pop();
                pop++;
            }
        }
        while (!stk.empty() && stk.top() == popped[pop++]) stk.pop();
        return stk.empty();
    }
};
